patroni/liveness_check.py

```diff
@@ -0,0 +1,122 @@
+                self.cancellable._kill_process()
+            try:
+                with self.cancellable._lock:
+                    self.cancellable._start_process(self.cmd, env=self.env)
+                ret = self.cancellable._process.wait(timeout=self.config['timeout'])
+                logger.info("Liveness Probe Completed with status %s", ret)
+                if ret == 0:
+                    self.__lv_failures = 0
+                else:
+                    if self.config['max_failures'] > 0:
+                        self.__lv_failures += 1
+                    logger.error("Liveness probe failed. failures %s out of %s", self.__lv_failures, self.config[
+                        'max_failures'])
+            except TimeoutExpired:
+                self.cancellable._kill_process()
+                if self.config['max_failures'] > 0:
+                    self.__lv_failures += 1
+                logger.error("Liveness probe timedout. failures %s out of %s", self.__lv_failures, self.config[
+                    'max_failures'])
+            except Exception as e:
+                logger.error("Exception during liveness probe")
+                logger.exception(e)
+            self._stopevent.wait(self.config['interval'])
+
+
+class Liveness(object):
+
+    def __init__(self, config):
+        self.config = config
+        self.livenesscheck = None
+        self._lock = Lock()
+
+    @synchronized
+    def reload_config(self, config):
+        self.config = config
+        if self.livenesscheck:
+            self.livenesscheck.config = self.config
+        # next loop iteration will restart if liveness checks are set in config
+        if self.livenesscheck and self.livenesscheck.is_alive():
+            self._disable()
+
+    def _disable(self):
+        if self.livenesscheck and self.livenesscheck.is_alive():
+            logger.info("liveness check alive, stopping")
+            self.livenesscheck.terminate()
+
+    def _activate(self, connect_args):
+        if not self.config:
+            logger.info("Liveness check activate skipped, No liveness checks in config")
+            return
+        self._disable()
+        if self.livenesscheck and self.livenesscheck.is_alive():
+            logger.info("Liveness check from prev run still active after terminate, skipping activate")
+            return
+        self.livenesscheck = LivenessThread(self.config, connect_args)
+        self.livenesscheck.start()
+
+    @synchronized
+    def disable(self):
+        self._disable()
+
+    @synchronized
+    def activate(self, connect_args):
+        self._activate(connect_args)
+
+    @property
+    def is_running(self):
+        return self.livenesscheck and self.livenesscheck.is_alive() or False
+
+    @property
+    def is_healthy(self):
+        if self.livenesscheck and self.livenesscheck.is_alive() and self.config['max_failures'] > 0:
+            return self.livenesscheck.lv_failures < self.config['max_failures']
+        return True
```

Comment:
Hi @ksarabu1, I am sorry that it took so long.

Please find attached [liveness_check.py](https://github.com/zalando/patroni/files/6330670/liveness_check.py.txt) draft.
The code is not tested, but it should give you some ideas:
- LivenessThread() is pretty much modeled like CancellableSubprocess() (call() became call_probe(), and cancel() became terminate()). I did it because CancellableSubprocess() is already well tested.
- The LivenessThread.run() became more compact
- LivenessThread() has its own reload_config() method (we need to update self.cmd)
- Liveness.is_running and Liveness.is_healthy properties are synchronized